### DISCUSSION TITLE
Trying to control a smart fan using tinytuya

### DISCUSSION THREAD

**Discussion by generalzu1** [jasonacox/tinytuya#477] [Q&A] [answered]
Please lmk if this is the right place to ask. 

I am trying to control a smart fan (this one) using a raspberry pi. I am able to control the operating mode (on or off) of the fan. using

```python

import tinytuya
from tinytuya import Contrib
import time

fan  = Contrib.ClimateDevice('dev_ID','IP_address','local_key',version = 3.3)
fan.turn_on()
time.sleep(3)
fan.turn_off()
```

This seems to work just fine. However, when we try to set the fan speed using `fan.set_fan_speed('middle')` but that doesn;t work for some reason. 

I tried to use `fan.get_fan_speed()` but the data returns `None`. It should return ` 'low' ` because that's the first setting. 

To localise the problem, I used ` fan.status_json() ` 

It returns : 

>return status[self.DPS_FAN]
> KeyError: '5'

The data that gets returned from the fan is:
`{'dps' : {'1': True , '2', 'Cool', '3': 1}}`

The function `status_json()` is included below:

```python

    def status_json(self):
        """Wrapper around status() that replace DPS indices with human readable labels."""
        status = self.status()["dps"]
        return {
            "Power On": status[self.DPS_POWER],
            "Set temperature": status[self.DPS_SET_TEMP],
            "Current temperature": status[self.DPS_CUR_TEMP],
            "Fan speed": status[self.DPS_FAN],
            "Temperature unit": status[self.DPS_TEMP_UNIT],
            "Sleep preset On": status[self.DPS_SLEEP_PRESET],
            "Swing On": status[self.DPS_SWING],
            "Operating mode": status[self.DPS_STATE],
            "Timer left": status[self.DPS_TIMER],
        }

```
Does anyone know why it only returns the first three but not the rest?

**Comment by uzlonewolf** [comment]
I believe `ClimateDevice` is designed for an air conditioner, not a fan, so the DPs it expects are going to be different.  I'd just use `OutletDevice` instead and pass in the values manually.  If you open devices.json in a text editor, what does it have in the `mapping` section for this device?

**Reply by generalzu1** [reply]
The `OutletDevice` doesn't have any functions to control the fan speed tho. The only other place I found anything about fan is the `ThermostatDevice`  but I don't think it's the right place. 

Could you tell me how to open devices.json in a text editor?

**Reply by jasonacox** [reply]
> Could you tell me how to open devices.json in a text editor?

What do you use to write your code?  You should be able to use that to read the devices.json file that was created when you ran the Tuya Wizard. You could also do this and search for your smart fan details:

```python
import json

# Open the JSON file
with open('devices.json', 'r') as f:
    # Load the JSON data
    data = json.load(f)

# Print the JSON data in a formatted way
print(json.dumps(data, indent=4))
```

Tuya devices use DPS values to control their behavior. `DeviceOutlet` can be used to read or write to them to change behavior.

**Reply by generalzu1** [reply]
I didn't have to run the wizard the first time I treid it. I just got the local key and device ID from the website through the API Explorer. 

So it turns out that the free-trial I was on just expired. I can't get the wizard to work. It's going to take some time to configure that before I can get the wizard runningDo you have any alternatives to bypass that? 

I already have the local key and the device ID from the fan.

**Reply by uzlonewolf** [reply, marked as answer]
It's going to require some trial and error, but you should be able to do it without running the wizard.

Create a monitor script using https://github.com/jasonacox/tinytuya/blob/master/examples/monitor.py and, with it running, change the speed and mode in the phone app.  You should see updates in the monitor script when you do that.  Once you have the values for the different speeds/modes you can put together a script to control it.  Something along the lines of:

```python
import tinytuya
import time

FAN_DP = 3       # CHANGE TO WHAT THE MONITOR SCRIPT SHOWS

FAN_SPEED_LOW = 1    # CHANGE TO WHAT THE MONITOR SCRIPT SHOWS
FAN_SPEED_MEDIUM = 2 # CHANGE TO WHAT THE MONITOR SCRIPT SHOWS
FAN_SPEED_HIGH = 3   # CHANGE TO WHAT THE MONITOR SCRIPT SHOWS


fan  = tinytuya.OutletDevice('dev_ID','IP_address','local_key',version = 3.3)
fan.turn_on()
time.sleep(3)
fan.set_value( FAN_DP, FAN_SPEED_MEDIUM )
time.sleep(3)
fan.turn_off()
```

**Reply by generalzu1** [reply]
Thanks for that. That makes a lot more sense. I will work on it and send an update soon.

**Reply by generalzu1** [reply]
So we were able to open a new acount, and we found this in the `devices.json` file:

```
{
        "name": "Window Fan (310)",
        "id": "ebde66bcd9c442fce1tgtk",
        "key": ">1/g3x|q0qJnrnLz",
        "mac": "d8:1f:12:19:f4:62",
        "uuid": "f3cff003a5cdd941",
        "sn": "10006121000C20",
        "category": "fs",
        "product_name": "Window Fan (310)",
        "product_id": "tqs2qloiq4hqiuso",
        "biz_type": 18,
        "model": "Comfort Zone Dual Window Fan",
        "sub": false,
        "icon": "https://images.tuyaus.com/smart/icon/bay1596489933724sjOD/35f62f21ab0910a27d31debe834a993e.png",
        "mapping": {
            "1": {
                "code": "switch",
                "type": "Boolean",
                "values": {}
            },
            "2": {
                "code": "mode",
                "type": "Enum",
                "values": {
                    "range": [
                        "Cool",
                        "Exhaust",
                        "Circulate"
                    ]
                }
            },
            "3": {
                "code": "fan_speed",
                "type": "Integer",
                "values": {
                    "unit": "",
                    "min": 1,
                    "max": 3,
                    "scale": 0,
                    "step": 1
                }
            }
        }
    }
```

I can see the code fan_speed, but I can't make sense of it. Could you guys explain if that means the fan_speed is being received correctly?

**Reply by generalzu1** [reply]
The monitor.py worked for us. Thanks a bunch!

**Reply by generalzu1** [reply]
I guiess we were just using the wrong python script the whole time.

**Reply by generalzu1** [reply]
I thought I will leave a note on what we were doing wrong. I was working on this with a friend.

We were just not able to interpret the datapoints correctly. I thought dp 5 belonged to the fan but it didn't. The function we were also using before was `set_fan_speed` which is not what we needed to use. 

It's only after I took a long look at the json file I posted above, it made sense that we were receiving the fan_speed correctly but we were using the wrong function and also the wrong datapoint. I believe it was mentioned above that the `ClimateDevice.py` where we got the function from, is actually for an air conditioner, but we were a little pig-headed and didn't know what else to do.

Then I understood the script that was provided by \u\uzlonewolf . I used that and it worked. 

Thanks again \u\uzlonewolf and \u\jasonacox !

**Reply by jasonacox** [reply]
Thanks for additional insight, @generalzu1 ! I'm glad you were able to figure it out. 👍
